### PR TITLE
Add basic doc to high level API.

### PR DESCRIPTION
**Pull request by bianpengyuan**
This adds basic doc to the high level API, so that user could have some reference to begin with. Instead of comment and document extensively on every method, this PR only adds minimum doc. We can reactively add more doc as needed. One part that remains to be done is stats API doc, since it is complicated and probably need some reorganization. Will do it in later PR.

Also this PR removes the old doc under `doc/` folder. It is hard to maintain and follow comparing with the code comment. SDK user should have no problem reading code and comment directly.

### CODE REVIEW

**Comment by bianpengyuan** [comment]
@esnible This is basic doc for the SDK repo. Might have some overlap with the comment that you are planning to add.

File: proxy_wasm_api.h [hunk 1]
```diff
@@ -749,7 +851,44 @@ inline WasmResult dequeueSharedQueue(uint32_t token, WasmDataPtr *data) {
   return result;
 }
 
-// Headers/Trailers
+/*************************
+    L7 Request API
```
Comment (PiotrSikora) [inline proxy_wasm_api.h:855]: `s/L7 Request/HTTP/`

File: proxy_wasm_api.h [hunk 2]
```diff
@@ -970,7 +1124,9 @@ inline WasmResult getMetric(uint32_t metric_id, uint64_t *value) {
   return proxy_get_metric(metric_id, value);
 }
 
-// Higher level metrics interface.
+/************************
+      Stats API
```
Comment (PiotrSikora) [inline proxy_wasm_api.h:1128]: `s/Stats/Metrics/`

File: proxy_wasm_api.h [hunk 3]
```diff
@@ -193,6 +229,15 @@ struct Tuple3Hash {
...
 using HeaderStringPairs = std::vector<std::pair<std::string, std::string>>;
 
+/***********************
+    gRPC Handle API
+************************/
+
+/**
+ * GrpcCallHandlerBase is the base class for gRPC unary call handler.
+ * This class should not be used directly, instead GrpcCallHandler should
+ * be used.
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply proxy_wasm_api.h]: Removed

File: proxy_wasm_api.h [hunk 4]
```diff
@@ -87,7 +114,16 @@ inline void logAbort(std::string_view logMessag) {
...
 #define LOG_CRITICAL(...) LOG(Critical, __VA_ARGS__)
 
-// Buffers coming into the WASM filter.
+/***********************
+      Data API
+************************/
+
+/**
+ * WasmData represents data blob copied between the host env and the Wasm VM.
+ * WasmData does not own the underlying data. The data is owned by
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply proxy_wasm_api.h]: Ah I see, thanks!

File: proxy_wasm_api.h [hunk 5]
```diff
@@ -54,6 +51,32 @@ class ProxyException : std::runtime_error {
 };
 #endif
 
+/**
+ * High level C++ API, includes the following categories:
+ * Logging API: print log at host.
+ * Data API: represent buffered data copied between the host and the Wasm VM.
+ * Context API: represent runtime root context and stream context.
+                Root context also includes API for gRPC call and http call.
+ * gRPC Hanlde API: represent a gRPC callout.
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply proxy_wasm_api.h]: HTTP callout API is part of root context. I added a line for it, and mentioned that it is part of root context.

**Comment by brozansk** [comment]
What's stopping You from merging this? Having an outdated docs/wasm_filter.md is confusing...

**Pull request closed without merging**